### ISSUE TITLE
Feature Request: Support an implicit `@ignore` tag

### ISSUE THREAD

**Issue by fourpastmidnight** [pester/Pester#1115]
## 1. General summary of the issue

Adding the tag `@ignore` to a feature or scenario does not cause Pester to skip the feature or scenario.

## 2. Describe Your Environment

Pester version     : 4.4.2 C:\Program Files\WindowsPowerShell\Modules\Pester\4.4.2\Pester.psd1
PowerShell version : 5.1.17134.165
OS version         : Microsoft Windows NT 10.0.17134.0

## 3. Expected Behavior

I expected any features or scenarios tagged with `@ignore` to be ignored.

## 4.Current Behavior

Even though a feature or scenario is tagged with `@ignore`, Pester still runs the associated feature/scenario.

## 5. Possible Solution

Some implementations of Cucumber, such as SpecFlow for .NET, recognize an implicit `@ignore` tag to quickly mark tests that should be ignored during a test run. This is convenient when you're developing the tests and, for one reason or another, need to ignore one or more scenarios or features without having to explicitly call, in the case of Pester, `Invoke-Gherkin -ExcludeTag @ignore`.

## 6. Context

I was simply trying to ignore a scenario that, at the moment, was not executing correctly while refactoring some of the scenario step code. I admit, some might find this a bit too convenient. But I thought I'd put in the feature request/idea and let the community debate the merits of the request (or lack thereof).

**Comment by Jaykul** [comment]
@nohwnd I agree this would be useful. Does RSpec Pester have a tag that's excluded by default?

Implementing this is just a matter of adding a default value to -ExcludeTag, I think.

**Comment by nohwnd** [comment]
@Jaykul Nope, there is no tag that is excluded by default.

**Comment by Jaykul** [comment]
I don't think we want a feature like this to only affect one runner...

**Comment by nohwnd** [comment]
Why not? To me it's just another feature specific thing. Like being able to parse specifications. RSpec Pester also cannot do that.

**Comment by fourpastmidnight** [comment]
It seems in RSpec, there's an implicit `skip` tag (either `:skip => true` or `@skip => true`). Just like with cucumber's `@ignore`, when invoking `rspec`, there's no need to specify `--tags ~skip` to have tests tagged with `:skip` be ignored.

I decided I wanted to see what the output from cucumber looked like when using `@ignore`. But lo and behold, there is no such default tag in Cucumber proper! No, instead, SpecFlow has a default `@ignore` tag. I find it to be a useful feature, nonetheless.

_If_ it's decided this proposed feature would be helpful/useful for Pester RSpec style tests as well as for Gherkin tests, do we stick with the respective test framework nomenclature for "ignored/skipped" tests?

* **Pros:** It keeps the nomenclature consistent for the respective framework: Rspec style tests vs. Gherkin style tests (albeit, this default is for SpecFlow, the .NET implementation of Cucumber)
* **Cons:** Within Pester at large, depending on which test style you use, you would need to use a different tag to accomplish the same thing

A third option that could act as a bridge is for Pester to recognize both `skip` and `ignore` tags for both RSpec-style and Gherkin-style tests and act accordingly.

I don't think any one of these options is any "harder" to implement than the other, so this is purely a judgment call.

**Comment by Jaykul** [comment]
Three reasons, @nohwnd :

First, it just makes sense to me. The core implementation of tags and which to run/skip is shared in the PesterState ... 

Second, the idea of a tag that's ignored by default is not a feature of the _syntax_, it's a feature of the test framework. As @fourpastmidnight discovered, different gherkin runners do different things: have you seen the `--wip` feature and `@wip` tag in Cucumber? It's TDD-licious.

Finally, I am still holding on to the idea that we should rename the current `Invoke-Pester` as `Invoke-PSpec` and write a new `Invoke-Pester` that is aware of the possibility of multiple test syntaxes and calls **each** of them. In that command, you want to specify the tags to run or skip _once_, and not have different default tags added to each runner...

Of course, users can obviously add their own with `$PSDefaultParameterValues` ...

P.S. I specifically want `Invoke-Pester` to run multiple test runners while keeping a shared `PesterState` so we get one test results file (and even codecoverage?) for all the tests that get run -- this would allow teams to use one syntax for blackbox BDD specs and another for white-box unit tests, but also choose to switch from one syntax to another without having to rewrite all the tests at once.

**Comment by nohwnd** [comment]
@Jaykul Those features of the test framework are already there, the `-Skip` is the `@ingore` tag, and `-Pending` is the `@wip` tag. Unlike these magical tags the parameters are discoverable and explicit. So to me this was about making the Gherkin part consistent with what you can do in other test runners, but I don't think adding this to the RSpec part makes sense. 

There are also tons of features that I would like, but the framework cannot just change all it's syntax between versions, there are users who actually use it. :) But if you find someone willing to pay me two days a week for working on Pester I can start splitting up the core of the framework right now 😉 

But yeah it would be nice to isolate the core, and add some extension points to the framework so it acts like a collection of independent pieces. But I can't come up with a good way of adding extensibility in PowerShell. How would you do it? Ideally it would stay easily debuggable, and people would be able to swap pieces of the framework with their own implementations -> Say replacing the screen output with doing callbacks on a graphical test runner. - I tried doing this few times, but I could not come up with a good way of providing the extensibility, the closest is probably making a module that defines functions with given parameters, but it seemed quite crude. Is there a better way?

**Comment by Jaykul** [comment]
Oh, I momentarily forgot about `-Skip` -- we even use that 😆 

Obviously a `@skip` or `@ignore` tag is the gherkin equivalent, so that's fine, and I'll look at doing it.

I **have** been thinking about how to extend modules in modules. I think what we want is a lot like event handlers, where other modules can register a handler, and there's a way for _users_ to wire up handlers and clear them out too ... and store the exact set and order that we want (in each project?).  I've been thinking about this for a few weeks and am ready to start experimenting, if you want to have a chat in #PoshCode channel 😉

**Comment by nohwnd** [comment]
There is also another aspect, how to deploy it in the end. On one hand we want to keep it compact, so people can get the whole thing via a single `Install-Module` call. And on the other we want to publish some additional functionality (most likely as functions) so they can author their own extensions while reusing our code, but _without_ polluting the api for the poeple who just want to use Pester. I can imagine doing it in two ways: 
- Making the core a new module that the extensions use (not sure how easy this would be able to share state)
- publishing a global object that provides all the functionality (kinda like jQuery in JS) - sharing state would be easy, but it's not very powershellish.

**Comment by fourpastmidnight** [comment]
So, I just pushed up the branch [skip-rest-of-scenario-steps-after-first-non-passing-step](https://github.com/fourpastmidnight/Pester/tree/skip-rest-of-scenario-steps-after-first-non-passing-step) on my fork of Pester. There were several things I did in this branch, which are tangentially related to this conversation, which is why I'm commenting on it. (BTW, this branch sits on top of the work done for PR #1124.) I would've just created a PR, but I didn't want to waste CI build resources on this at the moment--as there are two tests that need to be fixed up due to the changes in this branch.

1. I wanted more Cucumber-like output from Gherkin-style tests. So this branch contains a very early first crack at customizing the output for the Gherkin test output while leaving the Pester output alone. It definitely needs work pursuant to conversations we've had here.
2. To make the above possible, Gherkin-style tests now skip all steps after the first failing step in a scenario. After all, why bother running the rest of the scenario if a step has failed? The scenario has already failed. This leads nicely to the next thing.
3. Allow Gherkin-Style steps to explicitly set themselves as Pending through a new function: Set-PendingStep. Cucumber does this when it gives you a "suggested" step implementation for an undefined step.
    * As with failed steps, once a Pending step is encountered, the rest of the scenario steps are skipped and the scenario is counted as Pending.
4. Allow Gherkin-style tests to recognize and report Undefined steps.
    * In the current implementation, Pester counts these as inconclusive. I still count them as "inconclusive" but they are displayed as Undefined.
    * Also currently, Pester would execute other steps which come after an inconclusive step. Since this makes no sense, the rest of the steps are skipped and the scenario is counted as Undefined
        * NOTE: Actually, there is a TODO on this right now--technically, the rest of the steps are currently not skipped, but that will change in a later commit.

I'm unhappy with a few things: namely in `Output.ps1`, I basically copied the existing `Write-PesterResult` function to a `Write-GherkinResult` function and modified it for Gherkin. But pursuant to our conversation here, this is where that "extensibility" model we've been talking about would come in handy, if possible. And this is what I plan to investigate next after cleaning up a few things with the current branch state.

Once I get this branch into a better state, I'll open a PR. But for now, I just wanted to give you guys a "sneak-peek" at it, especially as I begin to look at ways to extend and allow for the extensibility of Pester.
